### PR TITLE
Add Poly1305 implementation

### PR DESCRIPTION
**Pull request by jonasschnelli**
This adds a currently unused Poly1305 implementation including test vectors from RFC7539.

Required for BIP151 (and related to #15512).

### CODE REVIEW

**Comment by practicalswift** [comment]
Intentional unsigned wraparound is triggered in `crypto/poly1305.cpp` (on at least L122 and L124). That is fine but should be documented in `test/sanitizer_suppressions/ubsan` :-)

**Comment by jonasschnelli** [comment]
Benchmark done on a aarch64 (`RK3328 Quad-Core ARM Cortex A53 64-Bit`)

```
CHACHA20_1MB, 5, 340, 13.3704, 0.00772624, 0.00806671, 0.00775259
CHACHA20_256BYTES, 5, 250000, 2.41802, 1.92206e-06, 1.95887e-06, 1.92964e-06
CHACHA20_64BYTES, 5, 500000, 1.28993, 5.07247e-07, 5.26875e-07, 5.14802e-07
HASH_1MB, 5, 340, 22.5945, 0.0128931, 0.0135617, 0.0132931
HASH_256BYTES, 5, 250000, 6.84513, 5.43544e-06, 5.54961e-06, 5.45136e-06
HASH_64BYTES, 5, 500000, 7.47419, 2.94502e-06, 3.04259e-06, 2.99531e-06
POLY1305_1MB, 5, 340, 10.2145, 0.00591661, 0.00609435, 0.00601109
POLY1305_256BYTES, 5, 250000, 1.88446, 1.48188e-06, 1.52889e-06, 1.52119e-06
POLY1305_64BYTES, 5, 500000, 1.08101, 4.25452e-07, 4.43012e-07, 4.29685e-07
```

This may be interpreted as a possible performance increase in processing packets with ChaCha2Poly1305 versus Hash256 as checksum (especially small messages).

**Comment by jonasschnelli** [comment]
Benchmark on a Intel i7 with SSE4  i7-8700 CPU @ 3.20GHz)

```
CHACHA20_1MB, 5, 340, 2.42871, 0.00142497, 0.0014349, 0.00142748
CHACHA20_256BYTES, 5, 250000, 0.443971, 3.53501e-07, 3.56836e-07, 3.54826e-07
CHACHA20_64BYTES, 5, 500000, 0.235031, 9.37462e-08, 9.42958e-08, 9.39654e-08
HASH_1MB, 5, 340, 3.98642, 0.00233798, 0.00234913, 0.00234636
HASH_256BYTES, 5, 250000, 1.12901, 8.97008e-07, 9.10387e-07, 9.03194e-07
HASH_64BYTES, 5, 500000, 1.18437, 4.70553e-07, 4.77644e-07, 4.72031e-07
POLY1305_1MB, 5, 340, 0.974462, 0.000562465, 0.000591879, 0.000565398
POLY1305_256BYTES, 5, 250000, 0.187587, 1.48138e-07, 1.51869e-07, 1.49319e-07
POLY1305_64BYTES, 5, 500000, 0.117305, 4.6148e-08, 4.74575e-08, 4.69618e-08
```

**Comment by laanwj** [comment]
> Intentional unsigned wraparound is triggered in crypto/poly1305.cpp

Right—since when is unsigned integer wraparound a problem? I thought there was only undefined behavior in signed wraparound.

**Comment by sipa** [comment]
Yes, unsigned overflow is well-defined. Is -fsanitize=undefined enabling that by default? I don't see anything in our configuration turning it on specifically.

**Comment by practicalswift** [comment]
@laanwj Unsigned integer wraparound is perfectly well-defined. Intentional unsigned integer wraparound is not problematic at all. Is anyone claiming otherwise? :-)

**Comment by sipa** [comment]
@practicalswift Do you have any idea why the sanitizer catches it in that case?

**Comment by practicalswift** [comment]
@sipa Yes, it is `-fsanitize=integer` catching it.

An argument to keep `-fsanitize=integer` enabled in Travis it that it makes people aware of *unintentional*  unsigned integer wraparound (a common source of bugs, but obviously _not_ UB).

**Intentional** unsigned integer wraparounds (such as in hashing code) can simply be documented as such by adding a line to `test/sanitizer_suppressions/ubsan`.

File: src/crypto/poly1305.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+// Based on the public domain implementation by Andrew Moon
+// poly1305-donna-unrolled.c from https://github.com/floodyberry/poly1305-donna
```
Comment (sipa) [inline src/crypto/poly1305.cpp]: I can't really find the code this is based on anymore (perhaps it was moved?). It seems identical to code here though: https://github.com/jhcloos/openssh-chacha-poly1305/blob/master/poly1305-donna-unrolled.c ?

Comment (jonasschnelli) [reply src/crypto/poly1305.cpp]: The code is taken from the OpenSSH project (correct link is https://github.com/openssh/openssh-portable/blob/90e51d672711c19a36573be1785caf35019ae7a8/poly1305.c). I kept the reference (see https://github.com/openssh/openssh-portable/blob/90e51d672711c19a36573be1785caf35019ae7a8/poly1305.c#L3).

**Comment by sipa** [comment]
utACK e3d8d0f7b3102ec1e6b787feaac7fe98dd5a7d7a, compared with the original implementation and compared the test vectors with the RFC.

**Comment by laanwj** [comment]
This is giving me new compile warnings:
```
In file included from /.../bitcoin/src/test/crypto_tests.cpp:7:0:
/.../bitcoin/src/crypto/poly1305.h:19:66: warning: ‘__bounded__’ attribute directive ignored [-Wattributes]
     __attribute__((__bounded__(__minbytes__, 4, POLY1305_KEYLEN)));
                                                                  ^
/.../bitcoin/src/crypto/poly1305.h:19:66: warning: ‘__bounded__’ attribute directive ignored [-Wattributes]
/.../bitcoin/src/crypto/poly1305.h:19:66: warning: ‘__bounded__’ attribute directive ignored [-Wattributes]
```

**Comment by jonasschnelli** [comment]
Removed the __bounded__ attribute (lets don't bother with activate them).

**Comment by laanwj** [comment]
thanks,
utACK 89d9d0d63eb920d3008ce87f14e7654b34366c97

**Pull request merged**